### PR TITLE
Fix long log line issue (#13333)

### PR DESCRIPTION
**Pull request by mwuertinger**
Replace the Scanner with a Reader in the copySrc implementation to be
more resilient in case a log line exceeds the buffer size.

The bufer size is set to be 1MiB which should be enough for most use
cases.

### CODE REVIEW

**Comment by unclejack** [comment]
Allocating 1 MB for this is way too much. This won't scale.

**Comment by mwuertinger** [comment]
How much would you recommend?

File: daemon/logger/copier.go [hunk 1]
```diff
@@ -40,14 +45,24 @@ func (c *Copier) Run() {
...
-	for scanner.Scan() {
-		if err := c.dst.Log(&Message{ContainerID: c.cid, Line: scanner.Bytes(), Source: name, Timestamp: time.Now().UTC()}); err != nil {
-			logrus.Errorf("Failed to log msg %q for logger %s: %s", scanner.Bytes(), c.dst.Name(), err)
+
+	reader := bufio.NewReaderSize(src, MaxLogLineLength)
+	line, isPrefix, err := reader.ReadLine()
+	for err != io.EOF {
+		if err != nil {
+			logrus.Errorf("Failed to read log line for container %s: %s", c.cid, err)
+			continue
```
Comment (phemmer) [inline daemon/logger/copier.go]: This is going to go into an infinite loop if `err != nil && err != io.EOF`.

File: daemon/logger/copier.go [hunk 2]
```diff
@@ -40,14 +45,24 @@ func (c *Copier) Run() {
...
+		if isPrefix {
+			logrus.Warnf("Log line exceeds buffer size for container %s", c.cid)
+		}
+
+		err := c.dst.Log(&Message{ContainerID: c.cid, Line: line, Source: name, Timestamp: time.Now().UTC()})
+		if err != nil {
+			logrus.Errorf("Failed to log msg for container %s:", c.cid, err)
+		}
+
+		line, isPrefix, err = reader.ReadLine()
```
Comment (phemmer) [inline daemon/logger/copier.go]: `err` was redeclared on line 60. This means that the assignment to `err` here is going to the locally-scoped one, which is not the same one as on line 51. Basically, another infinite loop.

You need to keep the syntax the original code had in it. `if err := foo(); err != nil { ... }`

File: daemon/logger/copier.go [hunk 3]
```diff
@@ -40,14 +45,24 @@ func (c *Copier) Run() {
...
-	}
-	if err := scanner.Err(); err != nil {
-		logrus.Errorf("Error scanning log stream: %s", err)
+		if isPrefix {
+			logrus.Warnf("Log line exceeds buffer size for container %s", c.cid)
+		}
+
+		err := c.dst.Log(&Message{ContainerID: c.cid, Line: line, Source: name, Timestamp: time.Now().UTC()})
+		if err != nil {
+			logrus.Errorf("Failed to log msg for container %s:", c.cid, err)
```
Comment (phemmer) [inline daemon/logger/copier.go]: You're passing 2 parameters (`c.cid`, and `err`) to the template, but have only one placeholder (`%s`).

File: daemon/logger/copier.go [hunk 4]
```diff
@@ -40,14 +45,24 @@ func (c *Copier) Run() {
...
+	line, isPrefix, err := reader.ReadLine()
+	for err != io.EOF {
+		if err != nil {
+			logrus.Errorf("Failed to read log line for container %s: %s", c.cid, err)
+			continue
 		}
-	}
-	if err := scanner.Err(); err != nil {
-		logrus.Errorf("Error scanning log stream: %s", err)
+		if isPrefix {
```
Comment (LK4D4) [inline daemon/logger/copier.go]: Shouldn't we call `ReadLine` until we get `isPrefix == false`?

File: daemon/logger/copier.go [hunk 5]
```diff
@@ -40,14 +45,24 @@ func (c *Copier) Run() {
...
 func (c *Copier) copySrc(name string, src io.Reader) {
 	defer c.copyJobs.Done()
-	scanner := bufio.NewScanner(src)
-	for scanner.Scan() {
-		if err := c.dst.Log(&Message{ContainerID: c.cid, Line: scanner.Bytes(), Source: name, Timestamp: time.Now().UTC()}); err != nil {
-			logrus.Errorf("Failed to log msg %q for logger %s: %s", scanner.Bytes(), c.dst.Name(), err)
+
+	reader := bufio.NewReaderSize(src, MaxLogLineLength)
+	line, isPrefix, err := reader.ReadLine()
+	for err != io.EOF {
```
Comment (LK4D4) [inline daemon/logger/copier.go]: I think you should reformulate this to dedent code.

**Comment by unclejack** [comment]
This PR is missing its DCO for the commits. The other problem is that the general approach doesn't seem that good. There are multiple problems with the approach. Overall, NOT LGTM.

**Comment by unclejack** [comment]
@mwuertinger After discussing with @LK4D4, I'll close this PR and submit a new one which takes a different approach to fix the problem. The reason behind this decision is that this part of the code is quite sensitive to performance issues, as all logging code.

Thank you for your contribution. We hope to see more PRs from you in the future.

**Comment by mwuertinger** [comment]
Alright, looking forward to your implementation!

FWIW, I've been running it with my patch for 16 hours now with several containers producing tons of output (one of the log files is now almost 1GB large) and I didn't encounter any performance or stability issues.

**Pull request closed without merging**